
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>The ZiBRA Pipeline &#8212; Zibra Pipeline  documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="#">Zibra Pipeline  documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="the-zibra-pipeline">
<span id="the-zibra-pipeline"></span><h1>The ZiBRA Pipeline<a class="headerlink" href="#the-zibra-pipeline" title="Permalink to this headline">¶</a></h1>
<p>During the Zika project we have developed a fully integrated sequencing pipeline which is described in our paper at Nature Protocols.</p>
<p>This page describes in more detail the bioinformatics pipeline from that paper.</p>
<div class="section" id="installing-docker">
<span id="installing-docker"></span><h2>Installing Docker<a class="headerlink" href="#installing-docker" title="Permalink to this headline">¶</a></h2>
<p>For Linux:</p>
<p>https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-16-04</p>
<p>For Windows:</p>
<p>Ideally Docker for Windows:</p>
<p>https://docs.docker.com/docker-for-windows/</p>
<p>Some users may need to use Docker Toolbox:</p>
<p>https://docs.docker.com/toolbox/toolbox_install_windows/</p>
<p>For Mac:</p>
<p>https://docs.docker.com/docker-for-mac/</p>
</div>
<div class="section" id="quick-start">
<span id="quick-start"></span><h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h2>
<div class="section" id="minion-pipeline">
<span id="minion-pipeline"></span><h3>MinION pipeline<a class="headerlink" href="#minion-pipeline" title="Permalink to this headline">¶</a></h3>
<p>Install the Docker container:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">docker</span> <span class="n">pull</span> <span class="n">zibra</span><span class="o">/</span><span class="n">zibra</span><span class="p">:</span><span class="n">latest</span>
  <span class="n">docker</span> <span class="n">run</span> <span class="o">-</span><span class="n">t</span> <span class="o">-</span><span class="n">i</span> <span class="n">zibra</span><span class="o">/</span><span class="n">zibra</span><span class="p">:</span><span class="n">latest</span> <span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">bash</span>
</pre></div>
</div>
<p>Download and run on the Zika test sample:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>  <span class="n">mkdir</span> <span class="n">whoref</span>
  <span class="n">cd</span> <span class="n">whoref</span>
  <span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">s3</span><span class="o">.</span><span class="n">climb</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">nanopore</span><span class="o">/</span><span class="n">Zika_Control_Material_R9</span><span class="o">.</span><span class="mi">4</span><span class="n">_2D</span><span class="o">.</span><span class="n">tar</span>
  <span class="n">tar</span> <span class="n">xvf</span> <span class="n">Zika_Control_Material_R9</span><span class="o">.</span><span class="mi">4</span><span class="n">_2D</span><span class="o">.</span><span class="n">tar</span>
  <span class="n">fast5_to_consensus</span><span class="o">.</span><span class="n">sh</span> <span class="n">ZikaAsian</span> <span class="n">WHO</span> <span class="mi">20161118</span><span class="n">_Zika</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="k">pass</span><span class="o">/</span><span class="n">NB08</span>
</pre></div>
</div>
<p>A number of files will be produced.</p>
<p>###&nbsp;Analysing diverse viral lineages</p>
<p>We designed this protocol in order to monitor evolutionary changes over short time-scales for example in outbreaks and epidemics that have been through a recent bottleneck. In practice it means this protocol has been mainly tested on viruses where sequeunces have &lt;1% sequence divergence from a reference. The pipeline relies on detecting differences from the reference genome, that in turn depends on the nanopolish software. Nanopolish works by testing all possible combinations of variants to determine which combination is best explained by the signal-level data produced by the nanopore device. Therefore, computational complexity rapidly increases as the sequence diverges (2^n, where n is the number of variants to test in a window).</p>
<p>By default the Zibra pipeline will test a maximum of 1,000,000 potential haplotypes in a window, which roughly equates to 2^20 mutations. This can be increased using the --max-haplotypes parameter to zibra.py, but at the expense of running time.</p>
</div>
<div class="section" id="local-offline-basecalling">
<span id="local-offline-basecalling"></span><h3>Local offline basecalling<a class="headerlink" href="#local-offline-basecalling" title="Permalink to this headline">¶</a></h3>
<p>Currently we are recommending Oxford Nanopore’s Albacore for local base calling. If basecalling offline on Windows or Mac, we recommend this is done on the native operating system rather than in the Docker container, for reasons of speed.</p>
<p>We also do not recommend using Albacore demultiplexing at present, as this is rather lenient as it requires only a single barcode copy.</p>
<p>To basecall an R9.4 1D dataset on Windows with eight threads use, if the reads are stored in <code class="docutils literal"><span class="pre">C:\data\reads\run</span></code> and you want the basecals to save to <code class="docutils literal"><span class="pre">C:\Users\nick\data\run</span></code> then run::</p>
<p><code class="docutils literal"><span class="pre">read_fast5_basecaller.py</span> <span class="pre">--input</span> <span class="pre">C:\data\reads\run</span> <span class="pre">--worker_threads</span> <span class="pre">8</span> <span class="pre">-c</span> <span class="pre">r94_450bps_linear.cfg</span> <span class="pre">-s</span> <span class="pre">C:\Users\nick\data\run</span> <span class="pre">-r</span> <span class="pre">-o</span> <span class="pre">fast5</span></code></p>
</div>
<div class="section" id="demultiplexing">
<span id="demultiplexing"></span><h3>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permalink to this headline">¶</a></h3>
<p>Demultiplexing (splitting reads by their barcodes) is currently rather convoluted in the current Zibra pipeline. Initially, reads are mapped against a reference genome and only mapped reads are considered further. These reads are then trimmed by considering their alignment coordinates and relating that to the known start and end coordinates of each amplicon (as determined by the specified Primal Scheme BED file), with 40 base flanking sequences retained.</p>
<p><img alt="Trimming" src="_images/trimming.png" /></p>
<p>Next, Porechop is run: the presence of double barcodes are enforced, with a minimum barcode identity of 80%. Because nanopolish cannot read trimmed files, the results of Porechop are used to reconstitute the original FASTA files but this time sorted into bins, and then the process is repeated to call consensus sequences.</p>
<p>The reason for this complicated process is to reduce any cross-barcode contamination caused through the presence of in silico chimeras, e.g. when the start or end of a read is erroneously detected. This is important for accurate determination of contamination in the negative control.</p>
</div>
<div class="section" id="mounting-a-local-directory">
<span id="mounting-a-local-directory"></span><h3>Mounting a local directory<a class="headerlink" href="#mounting-a-local-directory" title="Permalink to this headline">¶</a></h3>
<p>You will probably want to mount a local directory where your reads are kept.</p>
<p>On Windows, say my data is in <code class="docutils literal"><span class="pre">C:\Users\nick\data</span></code> you would run:</p>
<p>docker run -t -v //C/Users/nick/data:/data -i zibra/zibra:latest /bin/bash</p>
<p>On Mac:</p>
<p>docker run -t -v /Users/nick/data:/data -i zibra/zibra:latest /bin/bash</p>
<p>Then use Porechop (in the container) to demultiplex the reads:</p>
</div>
<div class="section" id="illumina-pipeline-quickstart">
<span id="illumina-pipeline-quickstart"></span><h3>Illumina pipeline- Quickstart<a class="headerlink" href="#illumina-pipeline-quickstart" title="Permalink to this headline">¶</a></h3>
<p>docker pull zibra/zibra:latest
docker run -t -i zibra/zibra:latest /bin/bashwget --no-check-certificate
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR512/007/SRR5122847/SRR5122847_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR512/007/SRR5122847/SRR5122847_2.fastq.gz
illumina_pipeline.sh SRR5122847 SRR5122847_1.fastq.gz SRR5122847_2.fastq.gz ZikaAsian</p>
</div>
</div>
<div class="section" id="credits">
<span id="credits"></span><h2>Credits<a class="headerlink" href="#credits" title="Permalink to this headline">¶</a></h2>
<p>The ZiBRA Pipeline was developed with contributions from:</p>
<ul class="simple">
<li>Nick Loman (MinION pipeline)</li>
<li>Jared Simpson (nanopolish SNP calling)</li>
<li>Matt Loose (nanopore demultiplexing script)</li>
<li>Karthik Gangavarapu (Illumina pipeline)</li>
<li>Nate Grubaugh (Illumina pipeline)</li>
<li>Kristian Andersen (Illumina pipeline)</li>
<li>Trevor Bedford (help with Docker and useful fixes)</li>
</ul>
<p>It relies on a whole heap of open source software, thank you to all contributors.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">The ZiBRA Pipeline</a><ul>
<li><a class="reference internal" href="#installing-docker">Installing Docker</a></li>
<li><a class="reference internal" href="#quick-start">Quick start</a><ul>
<li><a class="reference internal" href="#minion-pipeline">MinION pipeline</a></li>
<li><a class="reference internal" href="#local-offline-basecalling">Local offline basecalling</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a></li>
<li><a class="reference internal" href="#mounting-a-local-directory">Mounting a local directory</a></li>
<li><a class="reference internal" href="#illumina-pipeline-quickstart">Illumina pipeline- Quickstart</a></li>
</ul>
</li>
<li><a class="reference internal" href="#credits">Credits</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="#">Zibra Pipeline  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Nick Loman and the ZiBRA Project Consortium.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>